core/Pos/Util/Modifier.hs

```diff
@@ -161,6 +163,27 @@ delete
     => k -> MapModifier k v -> MapModifier k v
 delete k (MapModifier m) = MapModifier $ HM.insert k Nothing m
 
+data KeyState v
+    = KeyNotFound
+    | KeyDeleted
+    | KeyInserted v
+
+alter
+    :: (Eq k, Hashable k)
+    => (KeyState v -> KeyState v)
+    -> k
+    -> MapModifier k v
+    -> MapModifier k v
+alter f key (MapModifier mm) = MapModifier $ HM.alter transformedF key mm
```

Comment:
I would suggest giving `alter` the same type signature of `HM.alter`, for the _principle of least surprise_, as `alter` is defined (for `HashMap`) as:

```
(Eq k, Hashable k) => (Maybe v -> Maybe v) -> k -> HashMap k v -> HashMap k v
```

I think will also simplify your implementation, which is now required to do all this double wrapping.

Reply:
tbh I don't like types like `Maybe (Maybe v)`, it's not clear what it means. also `KeyState` describes a state of key very clearly.